File: apps/android/app/src/main/java/app/murmur/android/settings/Languages.kt
package app.murmur.android.settings

/**
 * Dictation languages: ISO-639-1 codes every supported speech provider accepts, with the English
 * name the formatting model is told about. Mirror of apps/desktop/src/shared/languages.ts; keep
 * the two tables identical so both apps offer the same choices and build the same prompt.
 */
object Languages {
    const val AUTO = "auto"
    const val AUTO_LABEL = "Auto-detect"

    val ALL: List<Pair<String, String>> = listOf(
        "af" to "Afrikaans",
        "ar" to "Arabic",
        "bg" to "Bulgarian",
        "ca" to "Catalan",
        "zh" to "Chinese",
        "hr" to "Croatian",
        "cs" to "Czech",
        "da" to "Danish",
        "nl" to "Dutch",
        "en" to "English",
        "et" to "Estonian",
        "fi" to "Finnish",
        "fr" to "French",
        "de" to "German",
        "el" to "Greek",
        "he" to "Hebrew",
        "hi" to "Hindi",
        "hu" to "Hungarian",
        "id" to "Indonesian",
        "it" to "Italian",
        "ja" to "Japanese",
        "ko" to "Korean",
        "lv" to "Latvian",
        "lt" to "Lithuanian",
        "ms" to "Malay",
        "no" to "Norwegian",
        "fa" to "Persian",
        "pl" to "Polish",
        "pt" to "Portuguese",
        "ro" to "Romanian",
        "ru" to "Russian",
        "sr" to "Serbian",
        "sk" to "Slovak",
        "sl" to "Slovenian",
        "es" to "Spanish",
        "sw" to "Swahili",
        "sv" to "Swedish",
        "ta" to "Tamil",
        "th" to "Thai",
        "tr" to "Turkish",
        "uk" to "Ukrainian",
        "ur" to "Urdu",
        "vi" to "Vietnamese"
    )

    /** What the picker shows: auto-detect first, then every language by name. */
    val OPTIONS: List<Pair<String, String>> = listOf(AUTO to AUTO_LABEL) + ALL

    private val byCode: Map<String, String> = ALL.toMap()

    /**
     * English name of a dictation language, or null for auto-detect, blanks and unknown codes.
     * Region subtags are ignored ("pt-BR" -> Portuguese) so a value synced from another client
     * still resolves.
     */
    fun name(code: String?): String? {
        if (code.isNullOrBlank()) return null
        val base = code.trim().lowercase().split('-', '_')[0]
        if (base.isEmpty() || base == AUTO) return null
        return byCode[base]
    }

    /** Picker label for a stored value; unknown codes are shown as-is rather than disappearing. */
    fun label(code: String?): String {
        if (code.isNullOrBlank() || code.trim().lowercase() == AUTO) return AUTO_LABEL
        return name(code) ?: code
    }
}

File: packages/text-engine/src/languages.ts
/**
 * Dictation languages. ISO-639-1 codes, which every supported speech provider accepts
 * (OpenAI-compatible `language`, Deepgram `language`, ElevenLabs `language_code`), paired with the
 * English name the formatting model is told about. The Android app carries the same table in
 * app/murmur/android/settings/Languages.kt; keep the two identical so both apps offer the same
 * choices and build the same prompt.
 */

export const AUTO_LANGUAGE = 'auto'

export interface Language {
  code: string
  name: string
}

export const LANGUAGES: readonly Language[] = [
  { code: 'af', name: 'Afrikaans' },
  { code: 'ar', name: 'Arabic' },
  { code: 'bg', name: 'Bulgarian' },
  { code: 'ca', name: 'Catalan' },
  { code: 'zh', name: 'Chinese' },
  { code: 'hr', name: 'Croatian' },
  { code: 'cs', name: 'Czech' },
  { code: 'da', name: 'Danish' },
  { code: 'nl', name: 'Dutch' },
  { code: 'en', name: 'English' },
  { code: 'et', name: 'Estonian' },
  { code: 'fi', name: 'Finnish' },
  { code: 'fr', name: 'French' },
  { code: 'de', name: 'German' },
  { code: 'el', name: 'Greek' },
  { code: 'he', name: 'Hebrew' },
  { code: 'hi', name: 'Hindi' },
  { code: 'hu', name: 'Hungarian' },
  { code: 'id', name: 'Indonesian' },
  { code: 'it', name: 'Italian' },
  { code: 'ja', name: 'Japanese' },
  { code: 'ko', name: 'Korean' },
  { code: 'lv', name: 'Latvian' },
  { code: 'lt', name: 'Lithuanian' },
  { code: 'ms', name: 'Malay' },
  { code: 'no', name: 'Norwegian' },
  { code: 'fa', name: 'Persian' },
  { code: 'pl', name: 'Polish' },
  { code: 'pt', name: 'Portuguese' },
  { code: 'ro', name: 'Romanian' },
  { code: 'ru', name: 'Russian' },
  { code: 'sr', name: 'Serbian' },
  { code: 'sk', name: 'Slovak' },
  { code: 'sl', name: 'Slovenian' },
  { code: 'es', name: 'Spanish' },
  { code: 'sw', name: 'Swahili' },
  { code: 'sv', name: 'Swedish' },
  { code: 'ta', name: 'Tamil' },
  { code: 'th', name: 'Thai' },
  { code: 'tr', name: 'Turkish' },
  { code: 'uk', name: 'Ukrainian' },
  { code: 'ur', name: 'Urdu' },
  { code: 'vi', name: 'Vietnamese' }
]

/** What the pickers show: auto-detect first, then every language by name. */
export const LANGUAGE_OPTIONS: readonly Language[] = [
  { code: AUTO_LANGUAGE, name: 'Auto-detect' },
  ...LANGUAGES
]

/**
 * English name of a dictation language, or undefined for auto-detect, blanks and codes we do not
 * know. Region subtags are ignored ("pt-BR" -> Portuguese) so a value synced from another client
 * still resolves.
 */
export function languageName(code: string | undefined | null): string | undefined {
  if (!code) return undefined
  const base = code.trim().toLowerCase().split(/[-_]/)[0]
  if (!base || base === AUTO_LANGUAGE) return undefined
  return LANGUAGES.find((l) => l.code === base)?.name
}

/** Picker label for a stored value; unknown codes are shown as-is rather than disappearing. */
export function languageLabel(code: string | undefined | null): string {
  if (!code || code.trim().toLowerCase() === AUTO_LANGUAGE) return 'Auto-detect'
  return languageName(code) ?? code
}
